WIP Added the Data.Pulse module

TimSheard opened the pull request:
A Pulse structure encapsulates a computation that can be broken into many discrete steps. A pulse structure is abstracted over a monad: m, and an answer type: t, so the concrete type of a pulse structure is written: (T m a).
 
The Pulseable class supplies operations used by the scheduling infrastructure to run a discrete set of steps with out knowing anything about the computation inside.  It does this by running: "pulse p" or "pulseM p", depending upon whether the computation is monadic or not.  There are currently two instances for folds (both left and right associative, monadic or not) over Lists and Maps.

TimSheard committed:
Added Data.Pulse to the semantics library.

TimSheard committed:
removed warnings

TimSheard committed:
Added the class MAccum in Data.Pulse to make first order pulsable data. Simple exampe in that file, and an exampke for dong the full reward calculation at the end of the file for Shelley.Spec.Ledger.Rewards.

TimSheard committed:
Integrated pulsing into the reward calculation.

TimSheard committed:
ormolised

Jimbo4350 commented on semantics/executable-spec/src/Data/Pulse.hs:
```diff
@@ -43,20 +53,25 @@ data PulseListM m ans where
+    where assoc LeftA = "left "
           assoc RightA = "right "
 
+isNil :: [a] -> Bool
```
Why not use `null`? https://hackage.haskell.org/package/base-4.14.1.0/docs/Prelude.html#v:null

TimSheard committed:
serial

TimSheard committed:
Broke createRupd into 3 phases.

TimSheard committed:
ormolise

TimSheard committed:
Added extra asc (ActiveSlotCoeff) arg to createRupd in Benchmarks

TimSheard committed:
completed breaking createRupd into 3 phases.

TimSheard committed:
Added Data.Closure, serializable functions.

TimSheard committed:
Added Data.Closure for serializable functions. Redid reward calculation this way. Replaces type LL with SLP (Serializable List based Pulser).

TimSheard committed:
Final cleanup, extra stuff to RewardSnapShot, defined PulsState.

TimSheard committed:
Introduced PulsingRewUpdate as a 3 constructor state: Waiting Pulsing Complete. Made that the State
in the RUpd rule. Introduced 3 operations startStep pulseStep completStep, used them to update the rule.
Fixed dependencies in all places in Shelley era (tests included). Some tests fail. But making everything
go in 1 rule succeeds.  liftSTS $ runProvM $ Complete <$> createRUpd slotsPerEpoch b es maxsupply asc
Must be something about cycling through the blocks I am not getting.

TimSheard committed:
Added completion to newEpochTransition

TimSheard committed:
Removed Waiting as constructor of PulsingRewUpdate, and use (StrictMaybe(PulsingRewUpdate)) instead

TimSheard committed:
Mary era translations fixed.

TimSheard committed:
ormolised

JaredCorduan commented:
closing in favor of #2142
